File: templates/ruflet_flutter_template/apple_packages/ruflet_apple/Sources/RufletExtensions/RufletMap/Sources/PolylineLayer.swift
import RufletEngine
import SwiftUI

struct RufletPolylineDescriptor {
  let control: RufletControl
  let coordinates: [RufletMapCoordinate]
  let borderWidth: Double
  let borderColor: Color
  let color: Color
  let pattern: RufletStrokePattern
  let lineCap: CGLineCap
  let lineJoin: CGLineJoin
  let strokeWidth: Double
  let usesMeters: Bool
  let gradientStops: [Double]
  let gradientColors: [Color]
  let cullingMargin: Double
  let minimumHittableRadius: Double
  let simplificationTolerance: Double
}

@MainActor
func rufletPolylineDescriptors(_ layer: RufletControl) -> [RufletPolylineDescriptor] {
  let cullingMargin = layer.number("culling_margin", default: 10) ?? 10
  let minimumHittableRadius = layer.number("min_hittable_radius", default: 10) ?? 10
  let simplificationTolerance = layer.number("simplification_tolerance", default: 0.3) ?? 0.3
  return layer.children("polylines", visibleOnly: false).compactMap {
    polyline -> RufletPolylineDescriptor? in
    guard polyline.type == "PolylineMarker" else { return nil }
    let points = rufletMapCoordinates(polyline.value("coordinates"))
    guard points.count >= 2 else { return nil }
    polyline.notifyParent = true
    return RufletPolylineDescriptor(
      control: polyline,
      coordinates: points,
      borderWidth: polyline.number("border_stroke_width", default: 0) ?? 0,
      borderColor: rufletMapColor(polyline.string("border_color"), default: .yellow),
      color: rufletMapColor(polyline.string("color"), default: .yellow),
      pattern: rufletStrokePattern(polyline.value("stroke_pattern")),
      lineCap: rufletLineCap(polyline.string("stroke_cap")),
      lineJoin: rufletLineJoin(polyline.string("stroke_join")),
      strokeWidth: polyline.number("stroke_width", default: 1) ?? 1,
      usesMeters: polyline.boolean("use_stroke_width_in_meter", default: false),
      gradientStops: polyline.value("colors_stop")?.array?.compactMap(\.number) ?? [],
      gradientColors: polyline.value("gradient_colors")?.array?.compactMap {
        guard let value = $0.text else { return nil }
        return parseColor(value)
      } ?? [],
      cullingMargin: cullingMargin,
      minimumHittableRadius: minimumHittableRadius,
      simplificationTolerance: simplificationTolerance)
  }
}

struct PolylineLayerControl: View {
  @ObservedObject var control: RufletControl
  var body: some View { EmptyView() }
}

File: templates/ruflet_flutter_template/ruflet_packages/ruflet_map/lib/src/polyline_layer.dart
import 'package:ruflet/ruflet.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_map/flutter_map.dart';

import 'utils/map.dart';

class PolylineLayerControl extends StatelessWidget with RufletStoreMixin {
  final Control control;

  const PolylineLayerControl({super.key, required this.control});

  @override
  Widget build(BuildContext context) {
    debugPrint("PolylineLayerControl build: ${control.id}");

    var polylines = control
        .children("polylines")
        .where((c) => c.type == "PolylineMarker")
        .map((polyline) {
      polyline.notifyParent = true;
      return Polyline(
          borderStrokeWidth: polyline.getDouble("border_stroke_width", 0)!,
          borderColor: polyline.getColor(
              "border_color", context, const Color(0xFFFFEB3B))!,
          color: polyline.getColor("color", context, const Color(0xFFFFEB3B))!,
          pattern: parseStrokePattern(
              polyline.get("stroke_pattern"), const StrokePattern.solid())!,
          strokeCap: polyline.getStrokeCap("stroke_cap", StrokeCap.round)!,
          strokeJoin: polyline.getStrokeJoin("stroke_join", StrokeJoin.round)!,
          strokeWidth: polyline.getDouble("stroke_width", 1.0)!,
          useStrokeWidthInMeter:
              polyline.getBool("use_stroke_width_in_meter", false)!,
          colorsStop: polyline
              .get("colors_stop", [])!
              .map((e) => parseDouble(e))
              .nonNulls
              .toList(),
          gradientColors: polyline
              .get("gradient_colors", [])!
              .map((e) => parseColor(e, RufletStyleTheme.of(context)))
              .nonNulls
              .toList(),
          points: polyline
              .get("coordinates", [])!
              .map((c) => parseLatLng(c))
              .nonNulls
              .toList());
    }).toList();

    return PolylineLayer(
      polylines: polylines,
      cullingMargin: control.getDouble("culling_margin", 10.0)!,
      minimumHitbox: control.getDouble("min_hittable_radius", 10.0)!,
      simplificationTolerance:
          control.getDouble("simplification_tolerance", 0.3)!,
    );
  }
}
